### ISSUE TITLE
Use QR algorithm instead of Jacobi algorithm

### ISSUE THREAD

**Issue by AlexanderMath** [graphcore-research/pyscf-ipu#81]
Our current `ipu_eigh` uses the Jacobi algorithm. It is believed for other hw accelerators that the QR algorithm becomes faster than jacobi for larger `d>=512`. Since we are targeting this, and `jacobi` is caused a few issues, we may consider the QR algorithm. 

Main blocker: The QR algorithm uses several clever shift/switch tricks which are tricky to implement. This difficulty could be alleviated if we found a working implementation under OS license we could simply port to IPU.

**Comment by AlexanderMath** [comment]
Initial implementation 76M cycles. Aim for 1M cycles or so. Code on this branch https://github.com/graphcore-research/pyscf-ipu/tree/hessenberg

Note: Algorithm is almost identical to tesselate_ipu.linalg.qr, it just multiplies with another H from the other side. 

![image](https://github.com/graphcore-research/pyscf-ipu/assets/8614529/def36263-f6d0-4752-96ff-74439ae34415)

**Comment by AlexanderMath** [comment]
@balancap Do you have any pointers on hard parts?

@paolot-gc is looking at improving above profile. I'll take a look at https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.linalg.eigh_tridiagonal.html and https://github.com/tensorflow/tensorflow/blob/v2.13.0/tensorflow/python/ops/linalg/linalg_impl.py#L1232-L1588 during weekend.

**Comment by AlexanderMath** [comment]
Profile of a single iteration @balancap 
![image](https://github.com/graphcore-research/pyscf-ipu/assets/8614529/5ed451cf-836a-452d-abc4-7509d04251e2)

**Comment by AlexanderMath** [comment]
@balancap @paolot-gc Problem: For `M.shape=(1024,1024)` with `M.T=M` we want `eigh(M)`. We use the classic `hessenberg(M)=tri_diagonal` to turn problem into `eigh(tri_diagonal)`. Literature claims for `eigvals(tri_diagonal)` are easy and `eigvcects(tri_diagonal)` are hard (e.g. jax.lax.eigh_tridiagonal only only eigvals and not eigvects). 

Here's an algorithm (credit to ([fhvilshoj](https://github.com/fhvilshoj)): Replicate `tri_diagonal` onto every tile (it's 1024*3 float32's so ~12kb) and perform inverse power iteration on `tri_diagonal-eighval[i]*I` in parallel on all tiles.

**Comment by awf** [comment]
The above is called "Simultaneous Iteration" in https://courses.engr.illinois.edu/cs554/fa2015/notes/12_eigenvalue_8up.pdf.   You can't make the eigenvalue gap arbitrarily large if lambda_i = lambda_{i+1}, so in practice you can't make it arbitrailty large if they are close to equal.  

In general, I guess we should re-title this issue "Speed up eigh computation", and the first task is to gather potential implementation strategies, e.g. by just grabbing the slide headings from the lecture above.

As all of these approaches end up with provisos such as "Algorithm is complicated to implement and difficult questions of numerical stability, eigenvector orthogonality,
and load balancing must be addressed", it's probably a good idea to see if existing code such as https://en.wikipedia.org/wiki/ARPACK has been ported to e.g. numpy.

**Comment by AlexanderMath** [comment]
> You can't make the eigenvalue gap arbitrarily large if lambda_i = lambda_{i+1}, so in practice you can't make it arbitrailty large if they are close to equal.

Agree. 

> The above is called "Simultaneous Iteration" in https://courses.engr.illinois.edu/cs554/fa2015/notes/12_eigenvalue_8up.pdf. 

Do you mean "parallel inverse iteration"? Simultaneous iteration doesn't use inverse and it requires normalization (?orthogonalization of the q eigenvectors). The suggested strategy above doesn't require that. 

doesn't have inverse (and it 
![image](https://github.com/graphcore-research/pyscf-ipu/assets/8614529/7f196d7d-2bd8-46e4-9ebe-64feb94cdb34)
